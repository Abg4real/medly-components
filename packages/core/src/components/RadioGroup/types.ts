import { FontSizes, FontWeights, RadioSizes } from '@medly-components/theme';
import { WithThemeProp } from '@medly-components/utils';
import { LabelPositions } from '../Label/types';

export interface Props extends WithThemeProp {
    /** Function to be called on selecting any radio */
    onChange?: (value: string) => void;
    /** To be used to disable radio selection */
    disabled?: boolean;
    /** Default selected radio option */
    defaultChecked?: string;
    /** Label position */
    labelPosition?: LabelPositions;
    /** Label size */
    labelSize?: FontSizes;
    /** Label weight */
    labelWeight?: FontWeights;
    /** Label color */
    labelColor?: string;
    /** To be used when it is required in any form */
    required?: boolean;
    /** Radio group label */
    label?: string;
    /** Name to be given to all radio option as html attribute */
    name?: string;
    /** Radio size */
    size?: RadioSizes;
}
