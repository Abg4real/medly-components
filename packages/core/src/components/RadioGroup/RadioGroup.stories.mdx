import { RadioGroup } from './RadioGroup';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './RadioGroup.stories';

<Meta title="Core" component={RadioGroup} />

# RadioGroup

### Displays group of radio

<Preview withToolbar>
    <Story name="RadioGroup">{stories.Basic()}</Story>
</Preview>

### Props

<Props of={RadioGroup} />

### Theme

Radio theme will be used for radio group.
