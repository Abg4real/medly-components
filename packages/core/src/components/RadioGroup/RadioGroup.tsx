import { WithStyle } from '@medly-components/utils';
import React, { SFC, useCallback } from 'react';
import FieldWithLabel from '../FieldWithLabel';
import Radio from '../Radio';
import { Props } from './types';

export const RadioGroup: SFC<Props> & WithStyle = React.memo(props => {
    const {
        size,
        name,
        label,
        required,
        options,
        labelPosition,
        onChange,
        value,
        disabled,
        labelSize,
        fullWidth,
        labelWeight,
        labelColor
    } = props;

    const handleOnChange = useCallback(
        (event: React.FormEvent<HTMLInputElement>) => {
            const target = event.target as HTMLInputElement;
            onChange && onChange(target.value);
        },
        [onChange]
    );
    return (
        <FieldWithLabel fullWidth {...{ fullWidth, labelPosition }}>
            {label && (
                <FieldWithLabel.Label {...{ required, labelPosition, labelSize, labelWeight, labelColor }}>{label}</FieldWithLabel.Label>
            )}
            <FieldWithLabel.Field onChange={handleOnChange}>
                {options.map(option => (
                    <Radio
                        key={option.value}
                        {...{ ...option, size, name, required }}
                        disabled={disabled || option.disabled}
                        defaultChecked={option.value === value}
                    />
                ))}
            </FieldWithLabel.Field>
        </FieldWithLabel>
    );
});

RadioGroup.displayName = 'RadioGroup';
RadioGroup.Style = FieldWithLabel.Style;
RadioGroup.defaultProps = {
    label: '',
    name: '',
    fullWidth: true,
    labelPosition: 'left'
};
