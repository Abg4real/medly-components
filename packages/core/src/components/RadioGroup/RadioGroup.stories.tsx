import { boolean, select, text } from '@storybook/addon-knobs';
import React, { useState } from 'react';
import { RadioGroup } from './RadioGroup';
import { Props } from './types';

const labelPosition: Props['labelPosition'][] = ['left', 'right', 'top', 'bottom'];
const size: Props['size'][] = ['XS', 'S', 'M', 'L', 'XL'];

export const Basic = () => {
    const [value, setValue] = useState('female');
    return (
        <RadioGroup
            value={value}
            onChange={setValue}
            disabled={boolean('Disabled', false)}
            required={boolean('Required', false)}
            fullWidth={boolean('Full Width', false)}
            label={text('Label', 'Gender')}
            name={text('Name', 'gender')}
            size={select('Size', size, 'S')}
            labelPosition={select('Label Position', labelPosition, 'top')}
            options={[
                { value: 'female', label: 'Female' },
                { value: 'male', label: 'Male' }
            ]}
        />
    );
};
