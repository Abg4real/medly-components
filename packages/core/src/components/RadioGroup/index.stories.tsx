import { boolean, select, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import Radio from '../Radio';
import RadioGroup from './RadioGroup';
import { Props } from './types';

const labelPosition: Array<Props['labelPosition']> = ['left', 'right', 'top', 'bottom'];
const size: Array<Props['size']> = ['XS', 'S', 'M', 'L', 'XL'];

storiesOf('Core', module).add('RadioGroup', () => (
    <RadioGroup
        label={text('Label', 'Gender')}
        name={text('Name', 'gender')}
        required={boolean('Required', false)}
        size={select('Size', size, 'S')}
        labelPosition={select('Label Position', labelPosition, 'top')}
    >
        <Radio label="Female" />
        <Radio label="Male" />
    </RadioGroup>
));
