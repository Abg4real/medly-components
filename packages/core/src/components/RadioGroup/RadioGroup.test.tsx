import { fireEvent, render, screen } from '@test-utils';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { RadioGroup } from './RadioGroup';

const renderer = ({
    label = 'Gender',
    name = 'gender',
    helperText = 'Helper Text',
    errorText = undefined,
    onChange = jest.fn(),
    required = false,
    validator = undefined,
    onInvalid = jest.fn(),
    disabled = false,
    options = [
        { value: 'female', label: 'Female' },
        { value: 'male', label: 'Male' }
    ]
}) => render(<RadioGroup {...{ required, disabled, label, name, onInvalid, validator, helperText, options, errorText, onChange }} />);

describe('Radio Group', () => {
    it('should render properly', () => {
        const { container } = renderer({});
        expect(container).toMatchSnapshot();
    });

    it('should render disabled state properly', () => {
        const { container } = renderer({ disabled: true });
        expect(container).toMatchSnapshot();
    });

    it('should be able to render error text', () => {
        renderer({ errorText: 'Something went wrong' });
        const errorText = screen.getByText('Something went wrong');
        expect(errorText).toBeInTheDocument();
        expect(errorText).toHaveStyle(`color: rgb(204, 0, 0)`);
    });

    it('should call onChange prop with selected option', () => {
        const mockOnChange = jest.fn();
        renderer({ onChange: mockOnChange });
        userEvent.click(screen.getByText('Male'));
        expect(mockOnChange).toBeCalledWith('male');
    });

    it('should show the html5 form validation error', async () => {
        renderer({ required: true });
        fireEvent.invalid(screen.getByRole('radio', { name: 'Female' }));
        expect(screen.getByText('Constraints not satisfied')).toBeInTheDocument();
    });

    it('should call the validator function if provided', async () => {
        const mockOnInvalid = jest.fn();
        renderer({ onInvalid: mockOnInvalid, validator: (val: string) => (!val ? 'Please select any one gender' : '') });
        fireEvent.invalid(screen.getByRole('radio', { name: 'Female' }));
        expect(screen.getByText('Please select any one gender')).toBeInTheDocument();
        expect(mockOnInvalid).toBeCalled();
    });
});
