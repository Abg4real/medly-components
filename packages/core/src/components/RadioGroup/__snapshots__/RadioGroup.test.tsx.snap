// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio component should render correctly with all the default props 1`] = `
.c0 {
  margin: 8px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c1 {
  margin: 8px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c2 {
  margin: 0px 0px 0px 8px;
  color: #012040;
  cursor: pointer;
}

.c3 {
  width: 16px;
  height: 16px;
  position: relative;
  border-radius: 100%;
  border: 1px solid #CCCCCC;
  background-color: #ffffff;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  background-color: #012040;
  width: 0;
  height: 0;
  z-index: 1;
  border-radius: 100%;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  opacity: 0;
  z-index: 2;
  position: absolute;
  width: 100%;
  height: 100%;
  margin: 0;
  cursor: pointer;
}

.c4:focus {
  outline: none;
}

.c4:checked ~ .c5 {
  width: calc(75%);
  height: calc(75%);
  -webkit-transition: width 0.2s ease,height 0.2s ease;
  transition: width 0.2s ease,height 0.2s ease;
}

<div>
  <div
    class="c0"
  >
    
    <div
      class="c1"
    >
      <label
        class="c2"
        for="Female"
      >
        Female
      </label>
      <div
        class="c3"
      >
        <input
          class="c4"
          id="Female"
          name=""
          type="radio"
        />
        <div
          class="c5 c6"
        />
      </div>
    </div>
    <div
      class="c1"
    >
      <label
        class="c2"
        for="Male"
      >
        Male
      </label>
      <div
        class="c3"
      >
        <input
          class="c4"
          id="Male"
          name=""
          type="radio"
        />
        <div
          class="c5 c6"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Radio component should render correctly with all the props given 1`] = `
.c2 {
  margin: 8px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c0 {
  margin: 8px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  margin: 0px 0px 0px 8px;
  color: #012040;
  cursor: pointer;
}

.c1 {
  margin: 0px 0px 8px 0px;
  color: #012040;
}

.c1::after {
  color: red;
  content: ' *';
}

.c4 {
  width: 24px;
  height: 24px;
  position: relative;
  border-radius: 100%;
  border: 1px solid #CCCCCC;
  background-color: #ffffff;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  background-color: #012040;
  width: 0;
  height: 0;
  z-index: 1;
  border-radius: 100%;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  opacity: 0;
  z-index: 2;
  position: absolute;
  width: 100%;
  height: 100%;
  margin: 0;
  cursor: pointer;
}

.c5:focus {
  outline: none;
}

.c5:checked ~ .c6 {
  width: calc(75%);
  height: calc(75%);
  -webkit-transition: width 0.2s ease,height 0.2s ease;
  transition: width 0.2s ease,height 0.2s ease;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
      required=""
    >
      Gender
    </label>
    <div
      class="c2"
    >
      <label
        class="c3"
        for="Female"
      >
        Female
      </label>
      <div
        class="c4"
      >
        <input
          class="c5"
          id="Female"
          name="gender"
          required=""
          type="radio"
        />
        <div
          class="c6 c7"
        />
      </div>
    </div>
    <div
      class="c2"
    >
      <label
        class="c3"
        for="Male"
      >
        Male
      </label>
      <div
        class="c4"
      >
        <input
          class="c5"
          id="Male"
          name="gender"
          required=""
          type="radio"
        />
        <div
          class="c6 c7"
        />
      </div>
    </div>
  </div>
</div>
`;
